dotnet/src/webdriver/SeleniumManager.cs

```diff
@@ -0,0 +1,146 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Text;
+using System.Threading;
+using OpenQA.Selenium.Internal;
+
+namespace OpenQA.Selenium
+{
+    /// <summary>
+    /// Wrapper for the Selenium Manager binary.
+    /// </summary>
+    public static class SeleniumManager
+    {
+        private static string binary;
+        private static readonly List<string> KnownDrivers = new List<string>() {
+            "geckodriver.exe",
+            "chromedriver.exe"
+        };
+
+        /// <summary>
+        /// Determines the location of the correct driver.
+        /// </summary>
+        /// <param name="driverName">Which driver the service needs.</param>
+        /// <returns>
+        /// The location of the driver.
+        /// </returns>
+        public static string DriverPath(string driverName)
+        {
+            if (!KnownDrivers.Contains(driverName))
+            {
+                throw new WebDriverException("Unable to locate driver with name: " + driverName);
+            }
+            var binaryFile = Binary;
+            if (binaryFile == null) return null;
+
+            var arguments = "--driver " + driverName.Replace(".exe", "");
+            var output = RunCommand(binaryFile, arguments);
+            return output.Replace("INFO\t", "").TrimEnd();
+        }
+
+        /// <summary>
+        /// Gets the location of the correct Selenium Manager binary.
+        /// </summary>
+        private static string Binary
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(binary))
+                {
+                    string folder = "windows";
+                    string extension = ".exe";
+
+                    if (!Environment.OSVersion.Platform.ToString().StartsWith("Win"))
+                    {
+                        throw new WebDriverException("Selenium Manager only supports Windows in .NET at this time");
+                    }
+
+                    try
+                    {
+                        string name = "selenium-manager-" + folder;
+                        using (Stream fileStream = ResourceUtilities.GetResourceStream(name, name))
+                        {
+                            using (BinaryReader binReader = new BinaryReader(fileStream, Encoding.ASCII))
+                            {
+                                byte[] fileBytes = binReader.ReadBytes((int)fileStream.Length);
+                                string directoryName = string.Format(CultureInfo.InvariantCulture, "webdriver.{0}",
+                                    Guid.NewGuid().ToString("N"));
+                                var path = Path.Combine(Path.GetTempPath(), directoryName + "/" + folder);
```

Comment:
I don't think that writing to %temp% directory is good. Given that each time it creates new folder, disk space is used irrationally. Executing 100 tests produces 100 files which will be unpredictably cleaned up by OS. Supposably it's better to copy binary file to user's app space like `Environment.CurrentDirectory`, and make copying process thread-safe.

Reply:
well, it should be once per process, not once per test at least, since it is static method.
I think the next step is putting it in $Home/.cache/manager/ (similar to how we're storing the drivers themselves)
but then we'd need logic to make sure we have the latest version.